### PR TITLE
Spec and Boulder compatibility fixes

### PR DESCRIPTION
**Pull request by kuba**
Relevant acme-spec:
- letsencrypt/acme-spec#127
- letsencrypt/acme-spec#119
- letsencrypt/acme-spec#98
- letsencrypt/acme-spec#92

Relevant boulder:
- letsencrypt/boulder#170
- letsencrypt/boulder#128

### CODE REVIEW

File: acme/messages2_test.py [hunk 1]
```diff
@@ -32,15 +33,15 @@ def test_typ_prefix(self):
 
     def test_typ_decoder_missing_prefix(self):
         from acme.messages2 import Error
-        self.assertRaises(jose.DeserializationError, Error.from_json,
-                          {'detail': 'foo', 'type': 'malformed'})
-        self.assertRaises(jose.DeserializationError, Error.from_json,
-                          {'detail': 'foo', 'type': 'not valid bare type'})
+        self.jobj['type'] = 'malfomed'
+        self.assertRaises(jose.DeserializationError, Error.from_json, self.jobj)
+        self.jobj['type'] = 'not balid bare type'
```
Comment (jdkasten) [inline acme/messages2_test.py]: typo *valid

Comment (schoen) [inline acme/messages2_test.py]: Also "malformed" instead of "malfomed" (although it might well raise the same exception and pass the test!).

**Comment by jsha** [comment]
This looks good to me and fixes the client/server interop problems we saw.

**Comment by schoen** [comment]
I'll fix the typos after merging it.

**Comment by schoen** [comment]
I've fixed the typos on master.

**Pull request merged**